### PR TITLE
extend typing: base

### PR DESCRIPTION
**Pull request by Borda**
# Before submitting

- [x] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [ ] Did you make sure to update the docs?
- [ ] Did you write any new necessary tests?

## What does this PR do?
Fixes # (issue).

## PR review
Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: torchmetrics/metric.py [hunk 1]
```diff
@@ -481,107 +481,108 @@ def __hash__(self):
...
-    def __add__(self, other: Any):
+    def __add__(self, other: "Metric") -> "Metric":
         return CompositionalMetric(torch.add, self, other)
 
-    def __and__(self, other: Any):
+    def __and__(self, other: "Metric") -> "Metric":
         return CompositionalMetric(torch.bitwise_and, self, other)
 
-    def __eq__(self, other: Any):
+    # Fixme: this shall return bool instead of Metric
```
Comment (Borda) [inline torchmetrics/metric.py:490]: @justusschock any idea how we could fix this?

File: torchmetrics/metric.py [hunk 2]
```diff
@@ -92,20 +93,24 @@ def __init__(
...
         self._forward_cache = None
         self._update_called = False
 
         # initialize state
-        self._defaults = {}
-        self._persistent = {}
-        self._reductions = {}
+        self._defaults: dict = {}
+        self._persistent: dict = {}
+        self._reductions: dict = {}
```
Comment (justusschock) [inline torchmetrics/metric.py:105]:
```suggestion
        self._defaults: Dict[str, Union[List, Tensor]] = {}
        self._persistent: Dict[str, bool] = {}
        self._reductions: Dict[str, Union[str, Callable[[Union[List, Tensor]], Tensor]] = {}
```

File: torchmetrics/metric.py [hunk 3]
```diff
@@ -224,10 +229,10 @@ def _sync_dist(self, dist_sync_fn: Callable = gather_all_tensors, process_group:
             reduced = reduction_fn(output_dict[attr]) if reduction_fn is not None else output_dict[attr]
             setattr(self, attr, reduced)
 
-    def _wrap_update(self, update):
+    def _wrap_update(self, update: Callable) -> Callable:
 
         @functools.wraps(update)
-        def wrapped_func(*args, **kwargs):
+        def wrapped_func(*args: Tensor, **kwargs: Any) -> Optional[Any]:
```
Comment (justusschock) [inline torchmetrics/metric.py:235]:
```suggestion
        def wrapped_func(*args: Any, **kwargs: Any) -> Optional[Any]:
```

This probably has to match the signature of `update`

File: torchmetrics/metric.py [hunk 4]
```diff
@@ -305,10 +310,10 @@ def sync_context(
             for attr, val in cache.items():
                 setattr(self, attr, val)
 
-    def _wrap_compute(self, compute):
+    def _wrap_compute(self, compute: Callable) -> Callable:
 
         @functools.wraps(compute)
-        def wrapped_func(*args, **kwargs):
+        def wrapped_func(*args: Any, **kwargs: Any) -> Any:
```
Comment (justusschock) [inline torchmetrics/metric.py:316]:
```suggestion
        def wrapped_func() -> Any:
```

Can also match `compute` signature

File: torchmetrics/metric.py [hunk 5]
```diff
@@ -305,10 +310,10 @@ def sync_context(
             for attr, val in cache.items():
                 setattr(self, attr, val)
 
-    def _wrap_compute(self, compute):
+    def _wrap_compute(self, compute: Callable) -> Callable:
```
Comment (justusschock) [inline torchmetrics/metric.py:313]:
```suggestion
    def _wrap_compute(self, compute: Callable[Any]) -> Callable[Any]:
```

File: torchmetrics/metric.py [hunk 6]
```diff
@@ -450,7 +459,7 @@ def _load_from_state_dict(
             state_dict, prefix, local_metadata, True, missing_keys, unexpected_keys, error_msgs
         )
 
-    def _filter_kwargs(self, **kwargs):
+    def _filter_kwargs(self, **kwargs: Dict[str, Any]) -> Dict[str, Any]:
```
Comment (justusschock) [inline torchmetrics/metric.py:462]:
```suggestion
    def _filter_kwargs(self, **kwargs: Any) -> Dict[str, Any]:
```

kwargs are always of type Dict[str, X] so they are only annotated with X

File: torchmetrics/metric.py (hunk 1 above)
Comment (justusschock) [reply torchmetrics/metric.py:499]: I discussed with @tchaton  in the past. One can argue for bool and for metric (the metric checks the equality of the outputs for each argument then). Equality would probably go via `is` or hashes.

So the current behaviour at least is consistent with the other arithmetic and logic operators.

File: torchmetrics/collections.py [hunk 7]
```diff
@@ -207,7 +207,7 @@ def _to_renamed_ordered_dict(self) -> OrderedDict:
             od[self._set_name(k)] = v
         return od
 
-    def keys(self, keep_base: bool = False):
+    def keys(self, keep_base: bool = False) -> Iterable[str]:
```
Comment (justusschock) [inline torchmetrics/collections.py:210]:
```suggestion
    def keys(self, keep_base: bool = False) -> Iterable[Hashable]:
```

keys aren't necessarily strings

**Review by justusschock — APPROVED** [review]
I suggested some changes in typing. Maybe this means that we can get rid of the type ignores

File: torchmetrics/metric.py (hunk 1 above)
Comment (Borda) [reply torchmetrics/metric.py:499]: yep, I see the arithmetic but feel that logic like eq shall return a bool, that is also why mypy complains as we overwrite something which shall return bool but return metric...

Comment (justusschock) [reply torchmetrics/metric.py:499]: How strict are we with deprecations in metrics? I'm fine with changing it. But I think checking the equality of metrics is not that easy. Probably it would require some syncing and checking the states.

**Review by tchaton — APPROVED** [review]
LGTM !

File: torchmetrics/metric.py (hunk 1 above)
Comment (Borda) [reply torchmetrics/metric.py:499]: we keep 0.1 deprecation loop, but if needed we are fine with breaking changes

**Review by awaelchli — APPROVED** [review]
